### DISCUSSION TITLE
Patterns for updating "sync" atoms that depend on async atoms like in V1

### DISCUSSION THREAD

**Discussion by Pinpickle** [pmndrs/jotai#2003] [Q&A] [answered]
In my current project, (still on 1.X), I use this pattern where a "sync" atom depends on an async atom. It's not actually sync but once the async atom first loads it effectively is. Something like this:

```
const syncCountAtom = atom(0);
const asyncAddAtom = atom(Promise.resolve(1));
const countAtom = atom(
 (get) => {
    return get(asyncAddAtom) + get(syncCountAtom);
  },
  (get, set) => {
    set(syncCountAtom, (c) => c + 1);
  }
);
```

Once `countAtom` finishes suspending for the first time, I can set it and not worry about it ever suspending again.

In V2, there's no special suspense inside atoms, so async atoms are just promises. This means that my `countAtom` has to have an async getter. Whenever I set it, it suspends (for a _very_ small amount of time).

In some cases (not all), it's trivial to wrap in a React 18 `startTransition`, but I'm not sure what kind of guarantees I get when opting into concurrent features like that.

Here's a codesandbox showing what I mean:

https://codesandbox.io/s/sync-and-async-atom-setting-n4g46r?file=/src/App.tsx

Are there any patterns for getting the old behaviour back?

**Comment by dai-shi** [comment, marked as answer]
Thanks for opening the discussion.
I've considered a lot when introducing v2 breaking changes. Unfortunately, there are no silver bullets.
There are three ways to work around your case.

## Use `unstable_unwrap`

This is essentially to mitigate v1 usage. It's not one-to-one, but you can get sync atom from an async atom.
Hopefully, we'll soon remove `unstable_` prefix and document it.

## Jotai store follows React convention for promises.

It has `status` property and `value` property on fulfillment.
https://github.com/pmndrs/jotai/blob/2188d7557500e59c10415a9e74bb5cfc8a3f9c31/src/vanilla/store.ts#L38-L43
We don't know if we consider it a public API. It depends on how entire React community goes. For now, we should consider it an internal API.
Here's the example: https://codesandbox.io/s/sync-and-async-atom-setting-forked-p8q9zp

## useAtom's `delay` option

This is not documented yet as we have been unsure about the use case.
But, it's there for such cases like yours.
https://github.com/pmndrs/jotai/blob/2188d7557500e59c10415a9e74bb5cfc8a3f9c31/src/react/useAtomValue.ts#L44-L47
Here's the example: https://codesandbox.io/s/sync-and-async-atom-setting-forked-lsgwlz

**Reply by Pinpickle** [reply]
Not 1 but 3 options!

I know option 2 (using the React convention) is considered private right now but that seems like the most useful option for me -- it'd be great to make it public.

**Reply by Pinpickle** [reply]
@dai-shi I've been thinking this over since I asked about it, and I've come up with a PoC to implement the old suspense behaviour in Jotai 2:

```ts
type SuspenseGetter = <T,>(anAtom: Atom<T>) => Awaited<T>

type PromiseMeta<T> = {
  status?: 'pending' | 'fulfilled' | 'rejected'
  value?: T
  reason?: unknown
  orig?: PromiseLike<T>
}

const isPromise = <T,>(v: T | Promise<T>): v is Promise<T> & PromiseMeta<T> => {
  return v instanceof Promise;
}

const suspenseAtom = <T,>(readFn: (get: SuspenseGetter) => T) => {
  return atom(async (get) => {
    const getter = <T,>(anAtom: Atom<T | Promise<T>>): Awaited<Promise<T>> => {
      const value = get(anAtom);

      if (isPromise(value)) {
        if (value.status === "fulfilled") {
          return value.value as Awaited<T>;
        }

        if (value.status === "rejected") {
          throw value.reason
        }

        throw value;
      }

      return value as Awaited<T>;
    }

    const attemptRead = (): T | Promise<T> => {
      try {
        const v = readFn(getter);
        return v;
      } catch (e) {
        if (e instanceof Promise) {
          return e.then(
            () => attemptRead()
          )
        }

        throw e;
      }
    }

    return attemptRead();
  });
};
```

[Codesandbox here](https://codesandbox.io/s/mystifying-fast-48jdns?file=/src/App.tsx)

I imagine this would be more performant if implemented in core as you wouldn't need to wrap every atom like this. I'm also very sure there are thorny edge-cases this isn't accounting for.

This is essentially trying to use the semantics of the new `use` hook. I prefer this to `unstable_unwrap` as there's a useful path of what to do when the value isn't available. Maybe atoms could be passed a `getSync` method?

What do you think?

**Reply by dai-shi** [reply]
One of the reason we avoid it is to support Store API. Throwing a promise isn't good for the store API which has nothing to do with React.
Furthermore, in the future React version, throwing a promise isn't supported (probably warned or errored), and you need to use `use`.
So, I don't think it should be in core. A React-only util might be possible, but I would suggest to start it as a third-party library.

**Reply by Pinpickle** [reply]
@dai-shi I agree that `store.get` should not throw, even if atoms did support a suspense style thrown promise.

What I'm suggesting is providing a `getSync` or a `get.sync` function that allows atoms to opt into suspense-like behaviour in the context of its own read function. From outside the atom, including other atoms, it still exposes a normal value without any throw semantics.

So...

```
const myAsyncAtom = atom(Promise.resolve(0));

const suspendedAtom: Atom<number | Promise<number>> = atom((get) => {
  // get.sync might throw a promise but the atom handles this
  return get.sync(myAsyncAtom) + 1;
}); 

const value = await store.get(suspendedAtom) // this still works as expected
atom(async (get) => await suspendedAtom)) // so does this
```

My PoC has these semantics -- the return type of `suspenseAtom` is `Atom<T | Promise<T>>` and getting it will not throw.

**Reply by dai-shi** [reply]
Jotai core is framework agnostic since v2, and I believe the "throwing promise" behavior should never be in the core. We should use async/await which is the JS feature.

Our option is only a third-party package. Something like this might be possible.
```js
import { sync } from '...';
const suspendedAtom = atom((get) => {
  return sync(get)((get) => get(myAsyncAtom) + 1);
}); 
// Or maybe this:
const suspendedAtom = syncAtom((get) => {
  return get(myAsyncAtom) + 1;
}); 
```

> My PoC has these semantics -- the return type of suspenseAtom is Atom<T | Promise<T>> and getting it will not throw.

Just curious if it suspends in React?

**Reply by Pinpickle** [reply]
> Jotai core is framework agnostic since v2, and I believe the "throwing promise" behavior should never be in the core. We should use async/await which is the JS feature.

Makes sense that this is a design principle for Jotai now. It's a shame as throwing a promise is the only thing I can think of that allows computation to remain synchronous in the face of promises. It'd be so much better if JS allowed for custom promise runtimes 😢 

I'll probably end up writing a little util like you described when I manage to migrate to v2. There'll be a performance overhead of wrapping _every_ atom in the try/catch check but tbh I can't even quantify that easily.

> Just curious if it suspends in React?

Because its an atom that returns a promise, it suspends in React because `useAtom` suspends. The idea is that the suspense "runtime" inside the atoms is invisible and when accessed from the outside it's just promises.

**Reply by dai-shi** [reply]
So, is it only the concern with v2 migration from v1? Would you still like to write async atoms in the sync style even if you start a new project with v2?

**Reply by Pinpickle** [reply]
@dai-shi that's exactly the concern with migrating. If I were writing a project today from scratch, I'd want the "sync-in-async" style.

Not because I find it intuitive or anything... I would personally prefer to use async/await if the semantics worked in the context of a reactive application.

The _only_ motivating factor for me is merging sync and async state together. Like the `countAtom` example I gave, I want to update `syncCountAtom` without `countAtom` returning a promise (and needing to wait for a microtask and suspending the component). 

```ts
const syncCountAtom = atom(0);
const asyncAddAtom = atom(Promise.resolve(1));
const countAtom = atom(
 (get) => {
    return get(asyncAddAtom) + get(syncCountAtom);
  },
  (get, set) => {
    set(syncCountAtom, (c) => c + 1);
  }
);
```

This would be the same as doing the computation inside the component.

```ts
const asyncAddAtom = atom(Promise.resolve(1));
const syncCountAtom = atom(0);

const MyComponent = () => {
  const asyncAdd = useAtomValue(asyncAddAtom);
  const syncCount = useAtomValue(syncCountAtom);
  const count = syncCount + asyncAdd;
  // etc...
}
```

The only way for this to happen (as far as I'm aware) is if `countAtom` does not return a promise.

To me, suspense seems the most ergonomic way to do this. `unstable_unwrap` would definitely work, but it requires more boilerplate:

```ts
const syncCountAtom = atom(0);
const asyncAddAtom = atom(Promise.resolve(1));
const countAtom = atom(
 (get) => {
    const addValue = unstable_unwrap(get(asyncAddAtom));
    if (addValue == null) {
      // wait forever
      return new Promise(() => {});
      // alternatively: return null;
    }

    return addValue + get(syncCountAtom);
  },
  (get, set) => {
    set(syncCountAtom, (c) => c + 1);
  }
);
```

**Reply by dai-shi** [reply]
> unstable_unwrap would definitely work, but it requires more boilerplate:

I'm not sure if I understand your last example. You want to make `countAtom` not suspend at all, right?
How does this feel?
```js
const syncCountAtom = atom(0);
const asyncAddAtom = atom(Promise.resolve(1));
const defaultAddValue = (prev) => prev ?? 0;
const countAtom = atom(
 (get) => {
    const addValue = get(unstable_unwrap(asyncAddAtom, defaultAddValue));
    return addValue + get(syncCountAtom);
  },
  (get, set) => {
    set(syncCountAtom, (c) => c + 1);
  }
);
```

OR

```js
const syncCountAtom = atom(0);
const asyncAddAtom = atom(Promise.resolve(1));
const unwrappedAsyncAddAtom = unstable_unwrap(asyncAddAtom, (prev) => prev ?? 0);
const countAtom = atom(
 (get) => {
    return get(unwrappedAsyncAddAtom) + get(syncCountAtom);
  },
  (get, set) => {
    set(syncCountAtom, (c) => c + 1);
  }
);
```

If you want to initially suspend with `countAtom`, it's different.

**Reply by Pinpickle** [reply]
Oops I was completely misunderstanding how `unstable_unwrap` works - I've never actually used it. Thanks for correcting that,

What I'm aiming for is:

`countAtom` to return a promise when its waiting on external data
`countAtom` to return a value synchronously when the data is available

I'm less convinced than ever than suspense-in-atoms is the way to implement this. Maybe it's actually not too much boilerplate and doing an early return is easier to reason about. I'll attempt a migration to v2 _eventually_ and I should figure something out then.

**Reply by pabo** [reply]
I am learning jotai, and have no experience with v1. I'm working on a v2 project and have come to the same use case as @Pinpickle : "sync" atoms that depend on async atoms.

> Would you still like to write async atoms in the sync style even if you start a new project with v2?

**Does this imply there's a different approach that avoids this type of problem?**

## All atoms sync
One approach I have tried is having all atoms be sync (with sane default values that the components can render sanely, instead of a separate `Suspense` fallback UI) and then having an `init` function do my data fetching, and when the data fetching resolves, to synchronously set the atoms needed. 

**pros**: simple, less boiler plate
**cons**: doesn't use `Suspense`, need to implement async-ness *somewhere* which may get messy

```js
// store.ts
const wordsAtom = atom([] as string[]);
const chosenIndexAtom = atom(0);

const chosenWordAtom = atom(get => {
  const chosenIndex = get(chosenIndexAtom);
  const words = get(wordsAtom);
  return words[chosenIndex];
});

const initialize = () => {
  fetch("/wordlist").then((res) => res.json()).then(({ words }) => {
    store.set(wordsAtom, words);
  });
};

initialize();

// component.tsx
const Word = () => {
  const [chosenWord] = useAtom(chosenWordAtom);
  const [chosenIndex, setChosenIndex] = useAtom(chosenIndexAtom);

  const handleSomeEvent = () => {
    // pretend this does something to choose a new chosen word

    setChosenIndex(index);
  };

  return <div>{chosenWord}</div>;
};

```

## async atoms as needed

The other approach is to use async atoms for async data. Additionally any (otherwise sync) atom that references an async atom needs to be converted to async. I think of it as the async atoms infecting all the other atoms. Async creep. 

```js
// store.ts
const asyncWordsAtom = atom(
  fetch("/wordlist").then((res) => res.json())
);

const chosenIndexAtom = atom(0);

const chosenWordAtom = atom(async get => {                         // 1
  const chosenIndex = get(chosenIndexAtom);
  const words = await get(asyncWordsAtom);
  return words[chosenIndex];
});

// component.tsx
const Word = () => {
  const [chosenWord] = useAtom(chosenWordAtom, {delay: 0});        // 2
  const [chosenIndex, setChosenIndex] = useAtom(chosenIndexAtom);

  const handleSomeEvent = () => {
    // pretend this does something to choose a new chosen word

    startTransition(() => setChosenIndex(index));                  // 3
  }

  return (
    <Suspense fallback="...loading">
      <div>{chosenWord}</div>;
    </Suspense>
  );
};
```
If I use any async atom: 

1. every other atom that references it must also be async
2. every `useAtom` needs a `{delay: 0}` to prevent UI flashing during updates
3. every action that updates an atom needs to be wrapped in `startTransition`

**pros**: can use `Suspense`
**cons**: more boiler plate, have to keep track of whether an atom is sync or async, unsure of `startTransition` trade offs

Is this pretty much accurate? I'm just trying to make sure I am not missing any jotai concepts, and that I understand my options here.

**Reply by dai-shi** [reply]
Check out `unwrap` for another option. 

https://github.com/pmndrs/jotai/blob/f7d7f74eb64660421dfea78d2be1e47816ffd6be/docs/utilities/async.mdx#L84-L92

**Comment by iwoplaza** [comment]
@Pinpickle 
> What I'm aiming for is:
> `countAtom` to return a promise when its waiting on external data
> `countAtom` to return a value synchronously when the data is available

I recently wrote an implementation of a utility that creates an atom that:
- Suspends when first resolving the asynchronous value
- Stores the last computed value and updates itself in the background

Seems to be exactly what you were describing. It took a couple tries to implement, but it proved very useful in our project's codebase. I have a couple utils like that with interesting properties, might post them as topics here, maybe they will help others looking for these solutions.

**Reply by stevemolitor** [reply]
I'd like to see that.

I asked a somewhat similar question; see [here](https://github.com/pmndrs/jotai/discussions/2144#discussioncomment-7076588) for solution based on a suggestion from @dai-shi. I wanted to asynchronously fetch an atom value on initial load, but then make synchronous updates to the atom as the user makes edits. So the atom can be either a promise (the initial value) or a regular value (edited value). 

The solution was an atom that checks if the value is a promise in its read function. If it's a promise it resolve it first before returning it. This approach could probably be extracted into a reusable util as well. I may try that later if I start to see a pattern in our codebase.

**Reply by Pinpickle** [reply]
I'd be interested to see the utility, @iwoplaza!

**Reply by simon3495** [reply]
@iwoplaza Do you mind sharing it?
